pw-pbkdf2.c

```diff
@@ -229,6 +246,95 @@ static int pbkdf2_encrypt(
 						  dk.bv_len, iteration,
 						  salt.bv_len, (const uint8_t *) salt.bv_val,
 						  dk.bv_len, dk_value);
+#elif HAVE_MOZNSS
+	SECStatus rv;
+	rv = NSS_NoDB_Init(".");
+	if (rv != SECSuccess)
+ 	{
+    		fprintf(stderr, "NSS initialization failed (err %d)\n",
+        	PR_GetError());
+  	}
+        PK11SlotInfo *slot = NULL;
+	SECAlgorithmID * algid = NULL;
+	PRArenaPool *arena1;
+	PRArenaPool *arena2;
+	PRArenaPool *arena3;
+	arena1 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena2 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	arena3 = PORT_NewArena(DER_DEFAULT_CHUNKSIZE);
+	SECItem salt_moz={ siBuffer, NULL, 0};
+	SECItem pwitem={ siBuffer, NULL, 0};
+	slot=PK11_GetBestSlot(SEC_OID_PKCS5_PBKDF2, NULL);
+	if (slot == NULL)
+	{
+		printf("error initializing slot");
+	}
+
+        if(!ber_bvcmp(scheme, &pbkdf2_scheme)){
+                SECITEM_AllocItem(arena1, &salt_moz, (unsigned int) PBKDF2_SALT_SIZE);
+                salt_moz.data = memcpy(salt_moz.data , salt_value, PBKDF2_SALT_SIZE);
+                SECITEM_AllocItem(arena2, &pwitem, (unsigned int)passwd->bv_len);
+                pwitem.data= memcpy(pwitem.data, passwd->bv_val, passwd->bv_len );
+                algid=PK11_CreatePBEV2AlgorithmID(SEC_OID_PKCS5_PBKDF2, SEC_OID_HMAC_SHA1, SEC_OID_HMAC_SHA1, PBKDF2_SHA1_DK_SIZE, iteration, &salt_moz);
+                PK11SymKey *key = NULL;
+                key=PK11_PBEKeyGen(slot, algid, &pwitem, 0, NULL);
+                SECStatus rv2 = PK11_ExtractKeyValue(key);
+                SECItem *symkey;
+                SECITEM_AllocItem(arena3, symkey, PBKDF2_SHA1_DK_SIZE);
+                symkey=PK11_GetKeyData(key);
+                dk.bv_val=memcpy(dk.bv_val, symkey->data, symkey->len);
+                PK11_FreeSymKey(key);
+        }else if(!ber_bvcmp(scheme, &pbkdf2_sha1_scheme)){
```

Comment:
You should write:
~~~
if(!ber_bvcmp(scheme, &pbkdf2_scheme) || !ber_bvcmp(scheme, &pbkdf2_sha1_scheme)){
~~~
We should write the same in OpenSSL GnuTLS :p

Reply:
done!

Reply:
i suppose this should be applied in all checks of ber_bvcmp(scheme, &pbkdf2_sha1_scheme)

Reply:
did it for GnuTLS also